src/filesystem-upload-item-manager.js

```diff
@@ -0,0 +1,124 @@
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+import Path from 'path';
+
+import { normalizePath } from './utils';
+import FileSystemUploadDirectory from './filesystem-upload-directory';
+import FileSystemUploadAsset from './filesystem-upload-asset';
+import {
+    cleanseFolderName,
+    cleanseAssetName,
+    getItemManagerParent,
+} from './filesystem-upload-utils';
+
+/**
+ * Keeps track of FileSystemUploadDirectory and FileSystemUploadAsset
+ * instances that have been retrieved using the manager. Its primary
+ * purpose is to ensure that an instance for any given path is only
+ * created once, then reused thereafter. It will also automatically
+ * create necessary instances for parent directories up until the 
+ * manager's root path.
+ */
+export default class FileSystemUploadItemManager {
+    /**
+     * Constructs a new, empty instance of the manager that will use
+     * a the given information.
+     * @param {FileSystemUploadOptions} uploadOptions Will be given to
+     *  each FileSystemUploadDirectory instance that is created.
+     * @param {string} rootPath The top-most path that the manager
+     *  will track.
+     */
+    constructor(uploadOptions, rootPath, keepFlat = false) {
+        this.uploadOptions = uploadOptions;
+        this.directories = new Map();
+        this.assets = new Map();
+        this.rootPath = normalizePath(rootPath);
+        this.keepFlat = keepFlat;
+    }
+
+    /**
+     * Retrieves an instance of FileSystemUploadDirectory for a given local path.
+     * The method will cache the instance, if necessary, and will also ensure that
+     * all parent instances up to the manager's root are also cached.
+     * @param {string} localPath Directory path whose information will be used
+     *  to create the FileSystemUploadDirectory instance.
+     * @returns {Promise} Resolved with a FileSystemUploadDirectory instance
+     *  representing of the given directory path.
+     */
+    async getDirectory(localPath) {
+        const normalizedPath = normalizePath(localPath);
+        let parent = await getItemManagerParent(this, this.rootPath, localPath);
+
+        if (!this.directories.has(normalizedPath)) {
+            this.directories.set(normalizedPath,
+                new FileSystemUploadDirectory(
+                    this.uploadOptions,
+                    normalizedPath,
+                    await cleanseFolderName(this.uploadOptions, Path.basename(normalizedPath)),
```

Comment:
ok. I made the change, but curious - is there a technical reason for this, or is it just a personal preference?

Reply:
A bit of both.
I saw people getting confused by such syntax, because it's easy to forget to use the `await` keyword if they need to make changes and this wreaks the code havoc. (Also it makes it a bit easier to read).